core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -0,0 +1,379 @@
+       * <p>
+       * Stores that have this characteristic must override the {@link #purgeExpired()} method.
+       */
+      EXPIRATION
+   }
+
+   /**
+    *
+    * @param ctx initialization context used to initialize this store
+    * @return a stage that when complete signals taht this store has been successfully started
+    */
+   CompletionStage<Void> start(InitializationContext ctx);
+
+   /**
+    *
+    * @return a stage that when complete signals that this store has been stopped
+    */
+   CompletionStage<Void> stop();
+
+   /**
+    *
+    * @return
+    */
+   default Set<Characteristic> characteristics() {
+      return EnumSet.noneOf(Characteristic.class);
+   }
+
+   /**
+    * Provides a way for the store to communicate which media type it requires for its key arguments. Infinispan
+    * will invoke this method immediately after the store is started providing the media type that in memory objects are
+    * stored via {@code storageMediaType} as well as all the media types this running instance can support converting
+    * to. The store then must choose from any of the provided media types. Any methods that are invoked after that
+    * accept a key argument will be guaranteed to have this key be in the desired media type. Also any return values
+    * from this store should be in this media type
+    * <p>
+    * Note that choosing a media type other than the storage media type will incur the cost of converting the key
+    * to and from the storage media type for input and output parameters.
+    * @param storageMediaType how the key is stored in memory
+    * @param supportedMediaTypes what media types Infinispan can convert to automatically on behalf of the store
+    * @return the media type the store desires for keys to be in when invoked and will return in
+    */
+   default MediaType getKeyMediaType(MediaType storageMediaType, Set<MediaType> supportedMediaTypes) {
+      return storageMediaType;
+   }
+
+   /**
+    *
+    * @param storageMediaType
+    * @param supportedMediaTypes
+    * @return
+    */
+   default MediaType getValueMediaType(MediaType storageMediaType, Set<MediaType> supportedMediaTypes) {
+      return storageMediaType;
+   }
+
+   /**
+    *
+    * @return
+    */
+   default CompletionStage<Boolean> isAvailable() {
+      return CompletableFutures.completedTrue();
+   }
+
+   /**
+    *
+    * @param segment
+    * @param key
+    * @return
+    */
+   CompletionStage<MarshallableEntry<K, V>> loadEntry(int segment, Object key);
+
+   /**
+    *
+    * @param segment
+    * @param key
+    * @return
+    */
+   default CompletionStage<Boolean> containsKey(int segment, Object key) {
+      return loadEntry(segment, key)
+            .thenApply(Objects::nonNull);
+   }
+
+   /**
+    *
+    * @param segment
+    * @param entry
+    * @return
+    */
+   CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> entry);
```

Comment:
`writeEntry`? to be similar to `loadEntry`... same for `delete`

Reply:
I can agree about `writeEntry`, but unfortunately `deleteEntry` is a bit weirder as it is really more like `deleteEntryForKey`. But I guess I can append Entry just so the naming is the same.

I think I will instead just rename `loadEntry` to just `load`. This was another naming thing that cropped up when Ryan changed MarshallableEntry.

Reply:
+1 for just `load`. We had to rename to `loadEntry` because the return type changed from `MarshalledEntry -> MarshallableEntry`.